qiskit_ode/signals/signals.py

```diff
@@ -11,216 +11,262 @@
-    envelope function and a carrier frequency.
-    """
+    def __str__(self) -> str:
+        """Return string representation."""
+        if self.name is not None:
+            return str(self.name)
 
-    def __init__(
-        self,
-        envelope: Union[Callable, complex, float, int],
-        carrier_freq: float = 0.0,
-        phase: float = 0.0,
-        name: str = None,
-    ):
-        """
-        Initializes a signal given by an envelop and an optional carrier.
+        if self.is_constant:
+            return "Constant({})".format(str(self(0.0)))
 
-        Args:
-            envelope: Envelope function of the signal.
-            carrier_freq: Frequency of the carrier.
-            phase: The phase of the carrier.
-            name: name of signal.
-        """
-        super().__init__(name)
+        return "Signal(carrier_freq={freq}, phase={phase})".format(
+            freq=str(self.carrier_freq), phase=str(self.phase)
+        )
 
-        if isinstance(envelope, (float, int)):
-            envelope = complex(envelope)
+    def __add__(self, other: "Signal") -> "SignalSum":
+        return signal_add(self, other)
 
-        if isinstance(envelope, complex):
-            self.envelope = lambda t: envelope
-        else:
-            self.envelope = envelope
+    def __radd__(self, other: "Signal") -> "SignalSum":
+        return self.__add__(other)
 
-        self._carrier_freq = Array(carrier_freq)
-        self._phase = Array(phase)
+    def __mul__(self, other: "Signal") -> "SignalSum":
+        return signal_multiply(self, other)
 
-    def envelope_value(self, t: float = 0.0) -> complex:
-        """Evaluates the envelope at time t."""
-        return self.envelope(t)
+    def __rmul__(self, other: "Signal") -> "SignalSum":
+        return self.__mul__(other)
 
-    def value(self, t: float = 0.0) -> Array:
-        """Return the value of the signal at time t."""
-        arg = 1j * 2 * np.pi * self.carrier_freq * t + 1j * self.phase
-        return np.real(self.envelope_value(t) * np.exp(arg))
+    def __neg__(self) -> "SignalSum":
+        return -1 * self
+
+    def __sub__(self, other: "Signal") -> "SignalSum":
+        return self + (-other)
+
+    def __rsub__(self, other: "Signal") -> "SignalSum":
+        return other + (-self)
 
     def conjugate(self):
-        """Return a new signal that is the complex conjugate of this one"""
-        return Signal(lambda t: self.envelope_value(t).conjugate(), -self.carrier_freq, -self.phase)
+        """Return a new signal whose complex value is the complex conjugate of this one."""
 
+        def conj_env(t):
+            return np.conjugate(self.envelope(t))
 
-class Constant(BaseSignal):
-    """
-    Constant signal that has no carrier and may appear in a model.
-    """
+        return Signal(conj_env, -self.carrier_freq, -self.phase)
+
+    def draw(
+        self,
+        t0: float,
+        tf: float,
+        n: int,
+        function: Optional[str] = "signal",
```

Comment:
This is a bit of a nit-pick but I'm not convinced by the name `function` here. I don't have a good alternative to offer either. Perhaps, `element` or `component`?

Reply:
Yeah I'm not crazy about the name `function` either, but I've also been unable to come up with a better alternative. I don't like `element` or `component` as I think there is a bit of naming collision with `SignalSum`/`DiscreteSignalSum`, which actually have elements/components. 

The only other alternative I can think of that seems okay is `method`. 

- Pro: this arg is literally choosing which method (in the class sense) that is being plotted 
- Con: "method" sounds like it's referring to a "plotting method"